packages/solana/src/providers/WalletConnectProvider.ts

```diff
@@ -0,0 +1,254 @@
+    return (
+      (this.session?.namespaces['solana']?.chains
+        ?.map(sessionChainId => {
+          // This is a workaround for wallets that only accept Solana deprecated networks
+          let chainId = sessionChainId
+          if (chainId === SolConstantsUtil.CHAIN_IDS.Deprecated_Mainnet) {
+            chainId = SolConstantsUtil.CHAIN_IDS.Mainnet
+          } else if (chainId === SolConstantsUtil.CHAIN_IDS.Deprecated_Devnet) {
+            chainId = SolConstantsUtil.CHAIN_IDS.Devnet
+          }
+
+          return this.requestedChains.find(chain => `solana:${chain.chainId}` === chainId)
+        })
+        .filter(Boolean) as Chain[]) || []
+    )
+  }
+
+  public get publicKey() {
+    const account = this.getAccount(false)
+
+    if (account) {
+      return new PublicKey(account.publicKey)
+    }
+
+    return undefined
+  }
+
+  public async connect() {
+    const rpcMap = this.requestedChains.reduce<Record<string, string>>((acc, chain) => {
+      acc[`solana:${chain.chainId}`] = chain.rpcUrl
+
+      return acc
+    }, {})
+
+    if (this.provider.session) {
+      this.session = this.provider.session
+    } else {
+      this.provider.on('display_uri', this.onUri)
+      this.session = await this.provider.connect({
+        optionalNamespaces: {
+          solana: {
+            chains: this.getRequestedChainsWithDeprecated(),
+            methods: [
+              'solana_signMessage',
+              'solana_signTransaction',
+              'solana_signAndSendTransaction'
+            ],
+            events: [],
+            rpcMap
+          }
+        }
+      })
+      this.provider.removeListener('display_uri', this.onUri)
+    }
+
+    const account = this.getAccount(true)
+
+    this.emit('connect', new PublicKey(account.publicKey))
+
+    return account.address
+  }
+
+  public async disconnect() {
+    await this.provider?.disconnect()
+    this.emit('disconnect', undefined)
+  }
+
+  public async signMessage(message: Uint8Array) {
+    const signedMessage = await this.request('solana_signMessage', {
+      message: base58.encode(message),
+      pubkey: this.getAccount(true).address
+    })
+
+    return base58.decode(signedMessage.signature)
+  }
+
+  public async signTransaction<T extends AnyTransaction>(transaction: T) {
+    const serializedTransaction = this.serializeTransaction(transaction)
+
+    const result = await this.request('solana_signTransaction', {
+      transaction: serializedTransaction,
+      pubkey: this.getAccount(true).address
+    })
+
+    const decodedTransaction = base58.decode(result.transaction)
+
+    if (isVersionedTransaction(transaction)) {
+      return VersionedTransaction.deserialize(decodedTransaction) as T
+    }
+
+    return Transaction.from(decodedTransaction) as T
+  }
+
+  public async signAndSendTransaction<T extends AnyTransaction>(
+    transaction: T,
+    sendOptions?: SendOptions
+  ) {
+    const serializedTransaction = this.serializeTransaction(transaction)
```

Comment:
Probably we would like to check if this method was approved in the session before calling it? To make sure the wallet does support it.

Reply:
We could fallback on sign maybe? I'm not sure...

Reply:
Is there a way to check if the wallet support the feature through the UniversalProvider?

Reply:
yes, `provider.session.namespaces` something like that